beacon-chain/utils/shuffle.go

```diff
@@ -80,62 +33,77 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
 // on the original implementation from protolambda, https://github.com/protolambda/eth2-shuffle
-//
+func ShuffledIndex(index uint64, indexCount uint64, seed [32]byte) (uint64, error) {
+	return innerShuffledIndex(index, indexCount, seed, true)
+}
+
+// UnShuffledIndex returns the inverse of ShuffledIndex. This implementation is based
+// on the original implementation from protolambda, https://github.com/protolambda/eth2-shuffle
+func UnShuffledIndex(index uint64, indexCount uint64, seed [32]byte) (uint64, error) {
+	return innerShuffledIndex(index, indexCount, seed, false)
+}
+
 // Spec pseudocode definition:
-// def get_permuted_index(index: int, list_size: int, seed: Bytes32) -> int:
+//   def get_shuffled_index(index: ValidatorIndex, index_count: int, seed: Bytes32) -> ValidatorIndex:
 //     """
-//     Return `p(index)` in a pseudorandom permutation `p` of `0...list_size - 1` with ``seed`` as entropy.
-//     Utilizes 'swap or not' shuffling found in
-//     https://link.springer.com/content/pdf/10.1007%2F978-3-642-32009-5_1.pdf
-//     See the 'generalized domain' algorithm on page 3.
+//     Return the shuffled validator index corresponding to ``seed`` (and ``index_count``).
 //     """
-//     assert index < list_size
-//     assert list_size <= 2**40
+//     assert index < index_count
+//     assert index_count <= 2**40
+//     # Swap or not (https://link.springer.com/content/pdf/10.1007%2F978-3-642-32009-5_1.pdf)
+//     # See the 'generalized domain' algorithm on page 3
 //     for round in range(SHUFFLE_ROUND_COUNT):
-//         pivot = bytes_to_int(hash(seed + int_to_bytes1(round))[0:8]) % list_size
-//         flip = (pivot - index) % list_size
+//         pivot = bytes_to_int(hash(seed + int_to_bytes(round, length=1))[0:8]) % index_count
+//         flip = (pivot + index_count - index) % index_count
 //         position = max(index, flip)
-//         source = hash(seed + int_to_bytes1(round) + int_to_bytes4(position // 256))
+//         source = hash(seed + int_to_bytes(round, length=1) + int_to_bytes(position // 256, length=4))
 //         byte = source[(position % 256) // 8]
 //         bit = (byte >> (position % 8)) % 2
 //         index = flip if bit else index
 //     return index
-func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error) {
+func innerShuffledIndex(index uint64, indexCount uint64, seed [32]byte, dir bool) (uint64, error) {
 	if params.BeaconConfig().ShuffleRoundCount == 0 {
 		return index, nil
 	}
-	if index >= listSize {
+	if index >= indexCount {
 		return 0, fmt.Errorf("input index %d out of bounds: %d",
-			index, listSize)
+			index, indexCount)
 	}
-	if listSize > maxShuffleListSize {
+	if indexCount > maxShuffleListSize {
 		return 0, fmt.Errorf("list size %d out of bounds",
-			listSize)
+			indexCount)
+	}
+	rounds := uint8(params.BeaconConfig().ShuffleRoundCount)
+	round := uint8(0)
+	if !dir {
+		// Start at last round.
```

Comment:
```suggestion
		// Starting last round and iterating through the rounds in reverse, un-swaps everything, effectively un-shuffling the list.
```

Reply:
These should be full sentences as Terence suggested. Please limit comment line length to 80 characters (break into a new line if comment is longer than 80 char). The latter isn't a hard rule, but helpful suggestion :)

Reply:
thanks, i hope my latest version is better